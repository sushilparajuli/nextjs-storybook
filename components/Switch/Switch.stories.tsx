import type { Meta, StoryObj } from "@storybook/react";
import { expect } from "@storybook/jest";
import { userEvent, within } from "@storybook/testing-library";

import { Switch } from "./Switch";

// More on how to set up stories at: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
const meta: Meta<typeof Switch> = {
  title: "Components/Switch",
  component: Switch,
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Switch>;

export const BasicSwitch: Story = {
  play: async ({ args,canvasElement }) => {
    const canvas = within(canvasElement);
    const switchElm = canvas.getByTestId("switch-label");
    await expect(switchElm).toBeInTheDocument();
    userEvent.click(switchElm);
    await expect(args.onChange).toHaveBeenCalled();
  }
};
