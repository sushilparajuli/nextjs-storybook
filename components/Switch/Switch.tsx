import styled from "@emotion/styled"
import { ChangeEvent, FC, useId } from "react"
import { Icon } from "../Icon";
import { boxShadow, transition } from "../styles";

type Props = {
    /** onChange callback */
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
}

const Wrapper = styled.label`
  & input {
    display: none;
  }
  & input:checked {
    & ~ label {
      background: ${({ theme }) => theme.components.primary};
      &::after {
        margin-left: 3.5rem;
        background: ${({ theme }) => theme.components.active};
        ${transition()};
      }
    }
  }
`;

const Label = styled.label`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
  height: 3rem;
  width: 6rem;
  border-radius: 1.6rem;
  background: ${({ theme }) => theme.components.background};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2)};
  &::after {
    content: "";
    margin-left: 0.5rem;
    width: 2.1rem;
    height: 2.1rem;
    border-radius: 50%;
    background: ${({ theme }) => theme.components.nonActive};
    ${transition()};
  }
`;

export const Switch : FC<Props> = ({onChange}) => {
  const fieldId = useId()
  return (
    <Wrapper>
      <input id={fieldId} type="checkbox" onChange={onChange} />
      <Label htmlFor={fieldId} data-testid="switch-label"></Label>
    </Wrapper>
  )
}