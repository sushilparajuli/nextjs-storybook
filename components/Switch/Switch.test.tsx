import { render, screen } from "@/test-utils";
import userEvent from "@testing-library/user-event";
import { Switch } from "./Switch";
describe("Switch test cases", () => {
  const onChange = jest.fn();

  it("should render", () => {
    const { asFragment } = render(<Switch onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("check onChange  callback", () => {
    render(<Switch onChange={onChange} />);
    const element = screen.getByTestId("switch-label");
    userEvent.click(element);
    expect(onChange).toHaveBeenCalled();
  });
});
