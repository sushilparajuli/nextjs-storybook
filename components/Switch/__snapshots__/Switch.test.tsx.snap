// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch test cases should render 1`] = `
<DocumentFragment>
  <label
    class="css-nnkgk3"
  >
    <input
      id=":r0:"
      type="checkbox"
    />
    <label
      class="css-sf5ln6"
      data-testid="switch-label"
      for=":r0:"
    />
  </label>
</DocumentFragment>
`;
